Funcky/Extensions/EnumerableExtensions/MaxOrNone.cs

```diff
@@ -0,0 +1,335 @@
+        /// <param name="source">A sequence of optional Decimal values to determine the maximum value of.</param>
+        /// <returns>The maximum value in the sequence or None.</returns>
+        [Pure]
+        public static Option<decimal> MaxOrNone(this IEnumerable<Option<decimal>> source)
+            => source.WhereSelect(Identity).MaxOrNone(Identity);
+
+        /// <summary>
+        /// Invokes a transform function on each element of a sequence and returns the maximum optional Decimal value. If the sequence is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values of TSource to determine the maximum value of.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The maximum value in the sequence or None.</returns>
+        [Pure]
+        public static Option<decimal> MaxOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, decimal> selector)
+            => source.Aggregate(Option<decimal>.None(), (max, current) => Option.Some(max.Match(selector(current), m => Max(selector(current), m))));
+
+        /// <summary>
+        /// Invokes a transform function on each element of a sequence and returns the maximum optional Decimal value. If the transforemd sequence only consists of none or is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values of TSource to determine the maximum value of.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The maximum value in the sequence or None.</returns>
+        [Pure]
+        public static Option<decimal> MaxOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, Option<decimal>> selector)
+            => source.WhereSelect(selector).MaxOrNone(Identity);
+
+        /// <summary>
+        /// Returns the maximum value in a sequence of generic values compared by a Comparer{T}. If the sequence is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of generic values of TSource to determine the maximum value of.</param>
+        /// <param name="ω">Dummy Parameter, do not use.</param>
+        /// <returns>The maximum value in the sequence or None.</returns>
+        [Pure]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0027:Public API with optional parameter(s) should have the most parameters amongst its public overloads.", Justification = "The RequireClass type is only for type deduction of the right overload.")]
+        public static Option<TSource> MaxOrNone<TSource>(this IEnumerable<TSource> source, RequireClass<TSource>? ω = null)
```

Comment:
Why are there separate overloads for `class` and `struct`? Both of these constraints imply `notnull` when nullable is enabled, why not have one overload with `where T : notnull` then? (This comment applies to `MinOrNone` too)

Reply:
Fix suggestion in #224

Reply:
I had compile issues and I could solve them that way.